### PR TITLE
[2019.x] File not installed if source used previously.

### PR DESCRIPTION
**Pull request by wxtim**
Address #2438 

- [x] New test created, which fails when put on 7.8.x

Response to bug reported by @jarich.

### CODE REVIEW

File: lib/python/rose/config_processors/fileinstall.py [hunk 1]
```diff
@@ -250,6 +250,19 @@ def _process(self, conf_tree, nodes, loc_dao, **kwargs):
...
+                if not target.is_out_of_date:
+                    conn = loc_dao.get_conn()
+                    prev_dep_locs = conn.execute(
+                        "SELECT * FROM dep_names WHERE name=?", [target.name]
+                    ).fetchall()
+                    prev_dep_locs = [i[1] for i in prev_dep_locs]
+                    prev_dep_locs = [loc_dao.select(i) for i in prev_dep_locs]
+                    if (
+                        [i.name for i in prev_dep_locs] !=
+                        [i.name for i in target.dep_locs]
```
Comment (oliver-sanders) [inline lib/python/rose/config_processors/fileinstall.py:263]: This ensures strict order is preserved, which may or may not be a good thing, will have to do some digging, my questions:

* Are `prev_dep_locs` always going to be stored in the same order as `target.dep_locs`?
  * I.E. Even if you don't change the `source` list could this equality resolve to `False` anyway.
* Does the order matter?

Comment (oliver-sanders) [reply lib/python/rose/config_processors/fileinstall.py:263]: Answer, yes the order does matter:

```
[file:multiple]
source=a b c
```

Would result in a file that looks like this:

```
cat a
cat b
cat c
```

Comment (wxtim) [reply lib/python/rose/config_processors/fileinstall.py:263]: I meant to sort, and forgot. Oops.

Comment (oliver-sanders) [reply lib/python/rose/config_processors/fileinstall.py:263]: No oops at all, I think the order can matter.

Comment (oliver-sanders) [reply lib/python/rose/config_processors/fileinstall.py:263]: After a bit of a play I'm happy that dep_names are stored in the right order too.

**Review by oliver-sanders — APPROVED** [review]

**Comment by wxtim** [comment]
Request @oliver-sanders  review change to testing this change.
Request @dpmatthews checks functionality.

File: lib/python/rose/config_processors/fileinstall.py (hunk 1 above)
Comment (wxtim) [reply lib/python/rose/config_processors/fileinstall.py:263]:
```suggestion
                        sorted([i.name for i in prev_dep_locs]) !=
                        sorted([i.name for i in target.dep_locs])
```

File: lib/python/rose/config_processors/fileinstall.py [hunk 2]
```diff
@@ -250,6 +250,19 @@ def _process(self, conf_tree, nodes, loc_dao, **kwargs):
                         if prev_path != path:
                             target.is_out_of_date = True
                             break
+                # See if any sources have changed names.
+                if not target.is_out_of_date:
+                    conn = loc_dao.get_conn()
```
Comment (oliver-sanders) [inline lib/python/rose/config_processors/fileinstall.py:255]: I think we should close this in a try/finally block.

**Review by oliver-sanders — APPROVED** [review]

File: lib/python/rose/config_processors/fileinstall.py (hunk 1 above)
Comment (oliver-sanders) [reply lib/python/rose/config_processors/fileinstall.py]: Agh, no, my comments here were backing up your previous decision *not* to sort.

The order of sources can effect the resulting output so exact (unsorted) equality must be tested.

File: lib/python/rose/config_processors/fileinstall.py [hunk 3]
```diff
@@ -250,6 +250,25 @@ def _process(self, conf_tree, nodes, loc_dao, **kwargs):
...
+                            "SELECT * FROM dep_names WHERE name=?",
+                            [target.name]
+                        ).fetchall()
+                        prev_dep_locs = [i[1] for i in prev_dep_locs]
+                        prev_dep_locs = [
+                            loc_dao.select(i) for i in prev_dep_locs
+                        ]
+                        if (
+                            sorted([i.name for i in prev_dep_locs]) !=
+                            sorted([i.name for i in target.dep_locs])
```
Comment (wxtim) [inline lib/python/rose/config_processors/fileinstall.py:267]:
```suggestion
                            [i.name for i in prev_dep_locs] !=
                            [i.name for i in target.dep_locs]
```

**Review by kinow — APPROVED** [review]

**Comment by kinow** [comment]
Not closing as `[2.0]` has a review pending (not sure if they need to go together, or if this one can be merged first). Thanks Tim!

**Pull request merged**